### ISSUE TITLE
Only one GPU during training

### ISSUE THREAD

**Issue by YinZhe0412** [mahmoodlab/CLAM#224]
Does CLAM only support one GPU during training and cannot utilize other GPUs? How can this be modified?

**Comment by fedshyvana** [comment]
I don't have explicit benchmark numbers but i did play with multi-gpu implementation e.g. using ddp but there was no speed benefit in my experience since most of the bottleneck is IO (from reading cached embeddings from SSD) while the work that the GPU does is very light. In fact in my experience there's not really even that big of a differnece in speed if you use CPU vs. GPU for training.

**Comment by YinZhe0412** [comment]
I noticed that the GPU memory usage is very low during model training. I'm not sure what could be causing this. Is it because of what you said that the GPU workload is light? Is there any way to speed up the process?
The training data consists of 1,400.
The code is CUDA_VISIBLE_DEVICES=0 python main.py --drop_out --early_stopping --lr 2e-4 --k 10 --label_frac 0.75 --exp_code task_1_tumor_vs_normal_75 --weighted_sample --bag_loss ce --inst_loss svm --task task_1_tumor_vs_normal --model_type clam_sb --log_data --data_root_dir BLSZC_FEATURES_DIRECTORY --csv_path Step_3.csv --results_dir results_duoxiancheng --splits_dir splits
![image](https://github.com/mahmoodlab/CLAM/assets/81342165/61e2f52b-10b9-4f6a-9c17-88f72e2b51c6)

**Comment by fedshyvana** [comment]
Hi yeah I think if it's bottlenecked by IO, just make sure that you're using store your features on fast NVME SSD instead of hard drives. Other than that, I am not super sure how you can speed things up further without changing the way you train the models (e.g. instead of using patches per slide, you can sample a maximum and use padding to pad multiple slides to the same length and train with batch size > 1).

**Comment by YinZhe0412** [comment]
Thank you for your careful answer.
